gobblin-temporal/src/main/java/org/apache/gobblin/temporal/workflows/service/ManagedWorkflowServiceStubs.java

```
@@ -0,0 +1,53 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.temporal.workflows.service;
+
+import java.io.Closeable;
+
+import io.temporal.serviceclient.WorkflowServiceStubs;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * A wrapper class of {@link WorkflowServiceStubs} that implements the Closeable interface.
+ * It manages the lifecycle of {@link WorkflowServiceStubs}, ensuring proper shutdown and resource cleanup.
+ */
+@Getter
+@Slf4j
+public class ManagedWorkflowServiceStubs implements Closeable {
+  private final WorkflowServiceStubs workflowServiceStubs;
+
+  public ManagedWorkflowServiceStubs(WorkflowServiceStubs serviceStubs) {
+    this.workflowServiceStubs = serviceStubs;
+  }
+
+  @Override
+  public void close() {
+    try {
+      try {
+        workflowServiceStubs.getOptions().getMetricsScope().close();
+      }
+      finally {
+        workflowServiceStubs.shutdown();
+      }
+    }
+    catch (Exception e) {
+      log.error("Exception occurred while closing ManagedWorkflowServiceStubs", e);
```

do we need to throw an exception here? what happens if workflowServiceStubs is not shutdown

---

No, swallowing the exception would be better choice since failure in closing metrics scope is not that critical.
Why would workflowServiceStubs not shutdown?

---

> closing metrics scope is not that critical

the only impact is that yarn AM would take longer to complete and not shutdown immediately, right?

---

I meant a **failure** in closing metrics scope is not that critical. Closing metrics scope is still critical because of the reason you mentioned.
Even if we throw an exception here, it will be caught [here](https://github.com/apache/gobblin/blob/7475e492be96c7ae4ebaab950478390e5a93a87b/gobblin-temporal/src/main/java/org/apache/gobblin/temporal/joblauncher/GobblinTemporalJobScheduler.java#L224) which will still not terminate all active threads on AM.
In case `MetricsScope.close` fails with an active thread, we can't do much there unless we make a change above.